### DISCUSSION TITLE
Support for aws lambda

### DISCUSSION THREAD

**Discussion by sachinmour** [uNetworking/uWebSockets.js#355] [Q&A] [answered]
This is how it could work
1. Lambda receives an event with details about http method, body, headers, cookies etc.
**2. Convert that event into a call that uWebSockets can accept.**
Currently there doesn't seem to be a way to do it. We would need a helper function from the core implementation that exposes this function.
I want to do this but i have no idea where to start. If anyone can help me, that'd be awesome.

**Comment by ghost** [comment]
Maybe I'm missing something but isn't the point of AWS lambda to not need a server? Why would you want to shove already parsed requests into a server?

**Comment by sachinmour** [comment]
You're right. But to allow the code to work with any setup and maintain consistency, people usually prefer it. For eg. https://github.com/awslabs/aws-serverless-express
Works exactly how i described above.

**Comment by hst-m** [comment]
Lambda is for functions run after an event to do some work and then close down, maximum exec time of 15 minutes, it only receives the initial event so you are only serving that 1 request, its not for putting a server in there, what you should be doing is auto-scaling EC2 instances, did you try that yet @sachinmour? I think [Elastic Beanstalk](https://aws.amazon.com/elasticbeanstalk) is what you are looking for EC2 auto-scale

**Comment by sachinmour** [comment]
@hst-m yes, i do know about Elastic Beanstalk and very much familiar with AWS infrastructures and i wouldn't use Elastic Beanstalk but much rather prefer Fargate if i had to.
Also, i don't think lambdas work the way you put it. Let me explain in brief-
1. Lambdas can serve only 1 request at a time.
Let's say it takes 100ms to complete a request and lambda timeout is 900 sec.
That means One lambda can serve around 9000 requests over 900 sec or about 10 request/sec.
The only gotcha here is these requests need to be made in a waterfall fashion (one after other).
2. If we were to send 100 requests a sec and it takes 1 sec to serve 1 request, then
it would spin 100 lambdas and serve all requests in 1 sec.
but instead if you limit to 1 lambda only, then it would serve 1 req/sec and finish in 100sec.

I hope that made it clear for you.

**Comment by hst-m** [comment]
Looks like Fargate does containers instead of full OS image. For Lambda yes that is how Lambda works, handles 1 event at a time, spins many for many events, so why are you trying to put a server inside a lambda function? Do you understand it makes no sense? Clears it up for you?

**Comment by sachinmour** [comment]
I'm not really sure how it doesn't make sense for you.
There are full scale libraries just for this use case used by thousands.
https://www.npmjs.com/package/apollo-server-lambda
https://github.com/awslabs/aws-serverless-express
https://github.com/dougmoscrop/serverless-http
I understand if you don't agree.

**Comment by sachinmour** [comment]
I do understand that the performance benefit that uWebSockets.js provides will all be pretty much lost inside lambda. This is a use case to increase adoption as people might wanna try it inside a lambda and see how it behaves and works before they jump on full board with it. This change will not give these users performance but a choice on how they use it.

**Comment by hst-m** [comment, marked as answer]
Nobody should be using uWS.js inside Lambda at all because it is a bad idea and doesnt work as a server, it cant connect with clients so no way to send out websockets etc, it is easy to put uWS.js in EC2 instance, you are focused on Lambda for some reason but you should be using EC2 or containers you could try

**Comment by hst-m** [comment]
You could use EC2 uWS.js server to call Lambda functions, but you should not be putting uWS.js inside Lambda, use it outside

**Comment by sachinmour** [comment]
Thank you for your thoughts @hst-m .
